packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,548 @@
+    isDisabled: !states.allowsSelection && !states.hasAction
+  });
+
+  let {isFocusVisible, focusProps} = useFocusRing();
+  let {checkboxProps} = useGridListSelectionCheckbox(
+    {key: item.key},
+    state
+  );
+
+  let props: TreeItemProps<unknown> = item.props;
+  let renderPropValues = React.useMemo(() => ({
+    ...states,
+    isHovered,
+    isFocusVisible,
+    isExpanded,
+    hasChildRows,
+    level,
+    selectionMode: state.selectionManager.selectionMode,
+    selectionBehavior: state.selectionManager.selectionBehavior
+  }), [states, isHovered, isFocusVisible, state.selectionManager, isExpanded, hasChildRows, level]);
+
+  let renderProps = useRenderProps({
+    ...props,
+    id: undefined,
+    children: item.rendered,
+    defaultClassName: 'react-aria-TreeItem',
+    values: renderPropValues
+  });
+
+  useEffect(() => {
+    if (!item.textValue) {
+      console.warn('A `textValue` prop is required for <TreeItem> elements in order to support accessibility features such as type to select.');
+    }
+  }, [item.textValue]);
+
+
+  let expandButtonProps = {
+    onPress: () => {
+      if (!states.isDisabled) {
+        state.toggleKey(item.key);
+      }
+    },
+    'aria-label': isExpanded ? stringFormatter.format('collapse') : stringFormatter.format('expand'),
+    // TODO: the below actually isn't enough to have keyboard navigation skip over it, we need it to be a span type button but
+    // RAC buttons are always a "button" type element.
+    tabIndex: isAndroid() ? -1 : null
+  };
+
+  let expandButtonRef = useRef<HTMLButtonElement>(null);
+  useEffect(() => {
+    if (hasChildRows && !expandButtonRef.current) {
+      console.warn('Expandable tree rows must contain a expand button so screen reader users can expand/collapse the row.');
+    }
+  // eslint-disable-next-line
+  }, []);
+
+  let children = useCachedChildren({
+    items: state.collection.getChildren!(item.key),
+    children: item => {
+      switch (item.type) {
+        case 'content': {
+          return <TreeRowContent values={renderPropValues} item={item} />;
+        }
+        // Skip item since we don't render the nested rows as children of the parent row, the flattened collection
+        // will render them each as siblings instead
+        case 'item':
+          return <></>;
+        default:
+          throw new Error('Unsupported element type in TreeRow: ' + item.type);
+      }
+    },
+    // TODO: double check if this is the best way to go about making sure TreeRowContent's render props is always up to date
+    dependencies: [renderPropValues]
+  });
+
+  return (
+    <>
+      <div
+        {...mergeProps(filterDOMProps(props as any), rowProps, focusProps, hoverProps)}
+        {...renderProps}
+        ref={ref}
+        // TODO: missing selectionBehavior, hasAction and allowsSelection data attribute equivalents (available in renderProps). Do we want those?
+        data-expanded={hasChildRows ? isExpanded : undefined}
```

Comment:
Table is similar here in that it doesn't have data attributes for the mentioned attributes which feels like it makes sense, not sure how helpful it would be to expose the above as data attributes